File: src/main/java/am/petstore/PetStore/user/service/DeviceService.java
package am.petstore.PetStore.user.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import am.petstore.PetStore.user.dao.DeviceDao;
import am.petstore.PetStore.user.entity.Device;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.*;

@Service
@Transactional
public class DeviceService {
    private final Logger log = LoggerFactory.getLogger(this.getClass().getName());
    private Map<Object, Object> data = new HashMap<>();
    private Map<Object, Object> model = new HashMap<>();

    private DeviceDao deviceDao;
    private ObjectMapper mapper;

    @PersistenceContext
    private EntityManager manager;

    @Autowired
    public DeviceService(DeviceDao deviceDao, ObjectMapper mapper) {
        this.deviceDao = deviceDao;
        this.mapper = mapper;
    }


    public ResponseEntity<Map<Object, Object>> addOrUpdateDevice(JsonNode jsonNode) {
        try {
            Device device = mapper.readValue(jsonNode.get("data").toString(), Device.class);
            String uid = device.getUid();
            String installDate = device.getFirst_install_date();
            log.info(device.toString());
            if (deviceDao.existsByUid(uid)) {
                deviceDao.updateDeviceByUid(uid, device.getApp_version(), device.getFirst_install_date(), device.getLanguage(),
                        device.getModel(), device.getPlatform(), device.getSdk_version(), device.getFirebase_token(), new Date());
                device = deviceDao.findByUid(uid);
                LoggerFactory.getLogger("UpdateDevice").info(device.toString());
                Set<String> installDates = device.getInstall_dates();
                boolean a = false;
                for (String install : installDates) {
                    if (install.equals(installDate)) {
                        a = true;
                        break;
                    }
                }
                if (!a) {
                    installDates.add(installDate);
                }

                device.setInstall_dates(installDates);
                manager.merge(device);
                return addOrUpdateDeviceOkResponse(device, "Updated");
            } else {
                Set<String> installDates = new HashSet<>();
                installDates.add(installDate);
                device.setInstall_dates(installDates);
                deviceDao.saveAndFlush(device);
                return addOrUpdateDeviceOkResponse(device, "Created");
            }

        } catch (Exception e) {
            e.printStackTrace();
            return badRequestResponse("Invalid data");
        }
    }

    public ResponseEntity<Map<Object, Object>> setDeviceFree(JsonNode jsonNode) {
        try {
            String uid = jsonNode.get("data").get("uid").asText();
            Device device = deviceDao.findByUid(uid);
            device.setUser(null);
            manager.merge(device);
            return addOrUpdateDeviceOkResponse(device,"Success");
        } catch (Exception e) {
            e.printStackTrace();
            return badRequestResponse("Invalid data");
        }
    }

    private ResponseEntity<Map<Object, Object>> badRequestResponse(String message) {
        data.clear();
        model.clear();
        model.put("code", 400);
        model.put("message", message);
        data.put("data", model);
        return ResponseEntity.badRequest().body(data);
    }

    private ResponseEntity<Map<Object, Object>> addOrUpdateDeviceOkResponse(Device device, String message) {
        data.clear();
        model.clear();
        model.put("code", 200);
        model.put("message", message);
        model.put("device", deviceDao.findByUid(device.getUid()));
        data.put("data", model);
        return ResponseEntity.ok(data);
    }
}

File: src/main/java/am/petstore/petstore/user/service/DeviceService.kt
package am.petstore.petstore.user.service

import am.petstore.petstore.user.dao.DeviceDao
import am.petstore.petstore.user.entity.Device
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*
import javax.persistence.EntityManager
import javax.persistence.EntityManagerFactory

@Service
@Transactional
class DeviceService @Autowired constructor(private val deviceDao: DeviceDao, private val mapper: ObjectMapper, private val entityManagerFactory: EntityManagerFactory) {
    private val log = LoggerFactory.getLogger(this.javaClass.name)
    private val data: MutableMap<Any, Any> = HashMap()
    private val model: MutableMap<Any, Any?> = HashMap()

    private lateinit var entityManager: EntityManager

    suspend fun addOrUpdateDevice(jsonNode: JsonNode): ResponseEntity<Map<Any, Any>> {
        return try {
            log.info(jsonNode.toString())
            var device = mapper.readValue(jsonNode["data"].toString(), Device::class.java)
            val uid = device.uid
            val installDate = device.firstInstallDate
            log.info(device.toString())
            if (deviceDao.existsByUid(uid)) {
                withContext(Dispatchers.Default) {
                    deviceDao.updateDeviceByUid(uid, device.appVersion, device.firstInstallDate, device.language,
                            device.model, device.platform, device.sdkVersion, device.firebaseToken, Date())
                }
                device = withContext(Dispatchers.Default) { deviceDao.findByUid(uid) }
                LoggerFactory.getLogger("UpdateDevice").info(device.toString())
                val installDates = device.installDates
                var a = false
                for (install in installDates!!) {
                    if (install == installDate) {
                        a = true
                        break
                    }
                }
                if (!a) {
                    installDates.add(installDate)
                }
                device.installDates = installDates
                withContext(Dispatchers.Default) {
                    entityManager = entityManagerFactory.createEntityManager()
                    entityManager.transaction.begin()
                    entityManager.merge(device)
                    entityManager.transaction.commit()
                }
                addOrUpdateDeviceOkResponse(device, "Updated")
            } else {
                val installDates: MutableSet<String?> = HashSet()
                installDates.add(installDate)
                device.installDates = installDates
                withContext(Dispatchers.Default) { deviceDao.saveAndFlush(device!!) }
                addOrUpdateDeviceOkResponse(device, "Created")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            badRequestResponse("Invalid data")
        }
    }

    suspend fun setDeviceFree(jsonNode: JsonNode): ResponseEntity<Map<Any, Any>> {
        return try {
            val uid = jsonNode["data"]["uid"].asText()
            val device = withContext(Dispatchers.Default) { deviceDao.findByUid(uid) }
            device?.user = null
            withContext(Dispatchers.Default) {
                entityManager = entityManagerFactory.createEntityManager()
                entityManager.transaction.begin()
                entityManager.merge(device)
            }
            addOrUpdateDeviceOkResponse(device, "Success")
        } catch (e: Exception) {
            e.printStackTrace()
            badRequestResponse("Invalid data")
        }
    }

    private fun badRequestResponse(message: String): ResponseEntity<Map<Any, Any>> {
        data.clear()
        model.clear()
        model["code"] = 400
        model["message"] = message
        data["data"] = model
        return ResponseEntity.badRequest().body(data)
    }

    private suspend fun addOrUpdateDeviceOkResponse(device: Device?, message: String): ResponseEntity<Map<Any, Any>> {
        data.clear()
        model.clear()
        model["code"] = 200
        model["message"] = message
        model["device"] = withContext(Dispatchers.Default) { deviceDao.findByUid(device?.uid) }
        data["data"] = model
        return ResponseEntity.ok(data)
    }

}